contracts/treasury/LivepeerGovernor.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/GovernorUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorVotesUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorVotesQuorumFractionUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorSettingsUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/governance/extensions/GovernorTimelockControlUpgradeable.sol";
+
+import "../bonding/libraries/EarningsPool.sol";
+import "../bonding/libraries/EarningsPoolLIP36.sol";
+import "../polling/PollCreator.sol";
+
+import "../ManagerProxyTarget.sol";
+import "../IController.sol";
+import "../rounds/IRoundsManager.sol";
+import "./GovernorCountingOverridable.sol";
+import "./BondingCheckpointsVotes.sol";
+
+/**
+ * @title LivepeerGovernor
+ * @notice Core contract for Livepeer governance, starting as the treasury governor.
+ * @dev If we ever add fields to this class or more extensions, make sure to add a storage gap to our custom
+ * GovernorCountingOverridable extension.
+ */
+contract LivepeerGovernor is
+    Initializable,
+    ManagerProxyTarget,
```

Comment:
Regarding storage gaps: I don't think we need it in the `ManaferProxyTarget` as those contracts are pretty much already tied to never change: all our existing contracts like `BondingManager`, `RoundsManager` etc already inherit from them. If we ever need to make changes to it we'll need to make a `V2` and only make new contracts use it instead.

I did wonder if we wanted to add a storage gap to `GovernorCountingOverridable` though, since if we add other contracts after it or any custom fields to `LivepeerGovernor`, we would be block from upgrading. We don't need it right now, but could be a safety measure for future upgrades so we don't need to remember it on code review. WDYT?

Reply:
OK I fixed the 2 storage layout issues on https://github.com/livepeer/protocol/pull/615/commits/0ad3a8fca7b121e82d67034cd92585967dc782a9

Specifically:
- Fix the inheritance order so that `ManagerProxyTarget` comes first, to guarantee storage layout compatibility with the proxy
- Add a storage `__gap` to `GovernorCountingOverridable` for safety in future upgrades

I also understood why nothing was exploding: due to C3-linearization of the parent contracts, the order in which the fields were laid out in storage was:
```
┌────────────────────────────────────────┬──────────────────────────────────┬──────────────┬────────┬
│                contract                │          state_variable          │ storage_slot │ offset │                         >
├────────────────────────────────────────┼──────────────────────────────────┼──────────────┼────────┼
│            LivepeerGovernor            │            controller            │      0       │   0    │                         >
│            LivepeerGovernor            │           _initialized           │      0       │   20   │                         >
│            LivepeerGovernor            │          _initializing           │      0       │   21   │                         >
│            LivepeerGovernor            │         targetContractId         │      1       │   0    │
```

Meaning a linearization as `Manager`, `Initializable`, `ManagerProxyTarget`, ...
The lucky strike on that order is that the fields from the `Initializable` parent (`_initialized` and `_initializing`) ended up using  a storage space that is comlpetely empty on the `ManagerProxy`, since they used the empty space after the `controller` field. So there was no conflict there, but out of luck and also not guaranteed by our code, so rather flimsy.

After re-ordering, the `Initializable` fields come only after the `targetContractId` from the `ManagerProxyTarget`, meaning 1 additional storage slot on the whole contract, which I think is worth it due to clearer storage layout guarantees.

Finally, I've also included a couple tests on that commit to exercise upgradability a bit, which was missing in the test package. Some lower level tests could probably be done using foundry, but I believe those are enough. Let me know if y'all think we need any more guarantees there apart from the tests we're gonna do in testnet+forked mainnet as well.

Reply:
Ha, interesting, yeah I see. So we stumbled into storage compatibility. Yeah, I think the extra storage is worth it for the upgradeability guarantees. Thanks for getting those tests in place, that's good to increase confidence.

Definitely something we should also have more experienced auditor(s) look at to make sure we're future-proof.